Fragments do not follow correct practices

Many fragments in the app do not have a default no-args constructor. This means the app will crash if the activity is stopped, removed from memory, and then resumed because the fragments cannot be recreated.

This happens often with the feature in #327 because the `FilterListActivity` is stopped due to high memory usage and `StashGridFragment` doesn't have a no-args constructor.

The solution is probably switching to using ViewModels, but there are lots of complex objects to figure out.

The stacktrace looks like this:
```
java.lang.RuntimeException: Unable to start activity ComponentInfo{com.github.damontecres.stashapp/com.github.damontecres.stashapp.FilterListActivity}: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment com.github.damontecres.stashapp.StashGridFragment: could not find Fragment constructor
    Caused by: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment com.github.damontecres.stashapp.StashGridFragment: could not find Fragment constructor
    Caused by: java.lang.NoSuchMethodException: com.github.damontecres.stashapp.StashGridFragment.<init> []
```

--- a/app/src/main/java/com/github/damontecres/stashapp/GalleryActivity.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/GalleryActivity.kt
@@ -144,6 +144,10 @@ class GalleryActivity : TabbedGridFragmentActivity(R.layout.gallery_activity) {
         ) {
             super.onViewCreated(view, savedInstanceState)
 
+            if (savedInstanceState != null) {
+                gallery = savedInstanceState.getParcelable("gallery")!!
+            }
+
             studioImage = view.findViewById(R.id.studio_image)
             ratingBar = view.findViewById(R.id.gallery_rating_bar)
             table = view.findViewById(R.id.gallery_table)
@@ -225,8 +229,8 @@ class GalleryActivity : TabbedGridFragmentActivity(R.layout.gallery_activity) {
                             photographer =
                                 Optional.present(
                                     StringCriterionInput(
-                                        galleryData.photographer!!,
-                                        CriterionModifier.EQUALS,
+                                        value = galleryData.photographer!!,
+                                        modifier = CriterionModifier.EQUALS,
                                     ),
                                 ),
                         )
@@ -245,6 +249,11 @@ class GalleryActivity : TabbedGridFragmentActivity(R.layout.gallery_activity) {
 
             table.addView(row)
         }
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putParcelable("gallery", gallery)
     }
 
     companion object {

--- a/app/src/main/java/com/github/damontecres/stashapp/ImageActivity.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/ImageActivity.kt
@@ -53,33 +53,31 @@ class ImageActivity : FragmentActivity(R.layout.activity_image) {
 
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        if (savedInstanceState == null) {
-            lifecycleScope.launch(StashCoroutineExceptionHandler()) {
-                val imageId = intent.getStringExtra(INTENT_IMAGE_ID)!!
+        lifecycleScope.launch(StashCoroutineExceptionHandler()) {
+            val imageId = intent.getStringExtra(INTENT_IMAGE_ID)!!
 
-                val queryEngine = QueryEngine(this@ImageActivity)
-                val image = queryEngine.getImage(imageId)!!
-                viewModel.setImage(image)
+            val queryEngine = QueryEngine(this@ImageActivity)
+            val image = queryEngine.getImage(imageId)!!
+            viewModel.setImage(image)
 
-                val pageSize =
-                    PreferenceManager.getDefaultSharedPreferences(this@ImageActivity)
-                        .getInt("maxSearchResults", 25)
+            val pageSize =
+                PreferenceManager.getDefaultSharedPreferences(this@ImageActivity)
+                    .getInt("maxSearchResults", 25)
 
-                currentPosition = intent.getIntExtra(INTENT_POSITION, -1)
+            currentPosition = intent.getIntExtra(INTENT_POSITION, -1)
 
-                val dataSupplier = createDataSupplier()
-                if (dataSupplier != null) {
-                    val pagingSource =
-                        StashPagingSource<FindImagesQuery.Data, ImageData, ImageData, CountImagesQuery.Data>(
-                            this@ImageActivity,
-                            pageSize,
-                            dataSupplier,
-                            useRandom = false,
-                        )
-                    pager = StashSparseFilterFetcher(pagingSource, pageSize)
-                    totalCount = pagingSource.getCount()
-                    canScrollImages = true
-                }
+            val dataSupplier = createDataSupplier()
+            if (dataSupplier != null) {
+                val pagingSource =
+                    StashPagingSource<FindImagesQuery.Data, ImageData, ImageData, CountImagesQuery.Data>(
+                        this@ImageActivity,
+                        pageSize,
+                        dataSupplier,
+                        useRandom = false,
+                    )
+                pager = StashSparseFilterFetcher(pagingSource, pageSize)
+                totalCount = pagingSource.getCount()
+                canScrollImages = true
             }
         }
 

--- a/app/src/main/java/com/github/damontecres/stashapp/MarkerActivity.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/MarkerActivity.kt
@@ -30,7 +30,9 @@ import androidx.leanback.widget.ListRow
 import androidx.leanback.widget.ListRowPresenter
 import androidx.leanback.widget.OnActionClickedListener
 import androidx.leanback.widget.SparseArrayObjectAdapter
+import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.ViewModel
 import androidx.lifecycle.lifecycleScope
 import com.bumptech.glide.request.target.CustomTarget
@@ -69,18 +71,23 @@ class MarkerActivity : FragmentActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_details)
         if (savedInstanceState == null) {
-            val marker = intent.getParcelableExtra<Marker>("marker")!!
             supportFragmentManager.beginTransaction()
-                .add(android.R.id.content, MarkerDetailsFragment(marker))
+                .add(android.R.id.content, MarkerDetailsFragment())
                 .commit()
         }
     }
 
-    class MarkerDetailsViewModel : ViewModel() {
+    class MarkerDetailsViewModel(private val state: SavedStateHandle) : ViewModel() {
         val seconds = MutableLiveData<Double>()
+
+        val marker: LiveData<Marker> = state.getLiveData("marker")
+
+        fun setMarker(newMarker: Marker) {
+            state["marker"] = newMarker
+        }
     }
 
-    class MarkerDetailsFragment(private var marker: Marker) : DetailsSupportFragment() {
+    class MarkerDetailsFragment : DetailsSupportFragment() {
         companion object {
             private const val TAG = "MarkerDetailsFragment"
 
@@ -105,8 +112,9 @@ class MarkerActivity : FragmentActivity() {
                 ArrayObjectAdapter(TagPresenter(PrimaryTagLongClickListener())),
             ) { tagIds ->
                 // Kind of hacky
+                val marker = viewModel.marker.value!!
                 val result = mutationEngine.setTagsOnMarker(marker.id, tagIds[1], marker.tagIds)
-                marker = marker.copy(primaryTagId = tagIds[1])
+                viewModel.setMarker(marker.copy(primaryTagId = tagIds[1]))
                 listOfNotNull(result?.primary_tag?.tagData)
             }
 
@@ -116,8 +124,9 @@ class MarkerActivity : FragmentActivity() {
                 ListRowManager.SparseArrayRowModifier(mAdapter, TAG_POS),
                 ArrayObjectAdapter(TagPresenter(TagLongClickListener())),
             ) { tagIds ->
+                val marker = viewModel.marker.value!!
                 val result = mutationEngine.setTagsOnMarker(marker.id, marker.primaryTagId, tagIds)
-                marker = marker.copy(tagIds = tagIds)
+                viewModel.setMarker(marker.copy(tagIds = tagIds))
                 result?.tags?.map { it.tagData }.orEmpty()
             }
 
@@ -160,6 +169,9 @@ class MarkerActivity : FragmentActivity() {
         override fun onCreate(savedInstanceState: Bundle?) {
             super.onCreate(savedInstanceState)
 
+            val marker = requireActivity().intent.getParcelableExtra<Marker>("marker")!!
+            viewModel.setMarker(marker)
+
             primaryTagRowManager.name = getString(R.string.stashapp_primary_tag)
 
             val lock = ReentrantReadWriteLock()
@@ -199,9 +211,10 @@ class MarkerActivity : FragmentActivity() {
         ) {
             super.onViewCreated(view, savedInstanceState)
             adapter = mAdapter
-            initializeBackground()
+            val marker = viewModel.marker.value!!
+            initializeBackground(marker)
             setupDetailsOverviewRowPresenter()
-            setupDetailsOverviewRow()
+            setupDetailsOverviewRow(marker)
 
             onItemViewClickedListener =
                 StashItemViewClickListener(requireContext(), actionClickListener)
@@ -214,6 +227,7 @@ class MarkerActivity : FragmentActivity() {
             )
 
             viewLifecycleOwner.lifecycleScope.launch(StashCoroutineExceptionHandler()) {
+                val marker = viewModel.marker.value!!
                 val sceneData = queryEngine.getScene(marker.sceneId)!!
                 val markerData = sceneData.scene_markers.first { it.id == marker.id }
 
@@ -245,7 +259,7 @@ class MarkerActivity : FragmentActivity() {
             mPresenterSelector.addClassPresenter(DetailsOverviewRow::class.java, detailsPresenter)
         }
 
-        private fun setupDetailsOverviewRow() {
+        private fun setupDetailsOverviewRow(marker: Marker) {
             val row = DetailsOverviewRow(marker)
 
             val screenshotUrl = marker.screenshot
@@ -300,7 +314,7 @@ class MarkerActivity : FragmentActivity() {
             mAdapter.set(1, row)
         }
 
-        private fun initializeBackground() {
+        private fun initializeBackground(marker: Marker) {
             if (mDetailsBackground.coverBitmap == null) {
                 mDetailsBackground.enableParallax()
 
@@ -336,7 +350,7 @@ class MarkerActivity : FragmentActivity() {
                     val id = data!!.getLongExtra(SearchForFragment.ID_KEY, -1)
                     if (id == StashAction.ADD_TAG.id) {
                         val tagId = data.getStringExtra(SearchForFragment.RESULT_ID_KEY)!!
-                        Log.d(TAG, "Adding tag $tagId to scene marker ${marker.id}")
+                        Log.d(TAG, "Adding tag $tagId to scene marker")
                         viewLifecycleOwner.lifecycleScope.launch(
                             StashCoroutineExceptionHandler { ex ->
                                 Toast.makeText(
@@ -357,7 +371,7 @@ class MarkerActivity : FragmentActivity() {
                         }
                     } else if (id == REPLACE_PRIMARY_ID) {
                         val tagId = data.getStringExtra(SearchForFragment.RESULT_ID_KEY)
-                        Log.d(TAG, "Setting primary tag to $tagId to scene marker ${marker.id}")
+                        Log.d(TAG, "Setting primary tag to $tagId to scene marker")
                         if (tagId != null) {
                             viewLifecycleOwner.lifecycleScope.launch(
                                 StashCoroutineExceptionHandler { ex ->

--- a/app/src/main/java/com/github/damontecres/stashapp/MovieFragment.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/MovieFragment.kt
@@ -1,6 +1,5 @@
 package com.github.damontecres.stashapp
 
-import android.content.Intent
 import android.os.Bundle
 import android.view.View
 import android.widget.ImageView
@@ -38,41 +37,36 @@ class MovieFragment : Fragment(R.layout.movie_view) {
 
         table = view.findViewById(R.id.movie_table)
 
-        val movie = requireActivity().intent.getParcelableExtra<Movie>("movie")
-        if (movie != null) {
-            titleText.text = movie.name
-            if (movie.frontImagePath != null) {
-                configureLayout(frontImage)
-                StashGlide.with(requireActivity(), movie.frontImagePath)
-                    .optionalCenterCrop()
-                    .error(StashPresenter.glideError(requireContext()))
-                    .into(frontImage)
-            }
-            if (movie.backImagePath != null) {
-                configureLayout(backImage)
-                StashGlide.with(requireActivity(), movie.backImagePath)
-                    .optionalCenterCrop()
-                    .error(StashPresenter.glideError(requireContext()))
-                    .into(backImage)
-            }
-            viewLifecycleOwner.lifecycleScope.launch(StashCoroutineExceptionHandler()) {
-                val queryEngine = QueryEngine(requireContext())
-                val movie = queryEngine.getMovie(movie.id)
-                addRow(
-                    R.string.stashapp_duration,
-                    movie?.duration?.toDuration(DurationUnit.MINUTES)?.toString(),
-                )
-                addRow(R.string.stashapp_date, movie?.date)
-                addRow(R.string.stashapp_studio, movie?.studio?.name)
-                addRow(R.string.stashapp_director, movie?.director)
-                addRow(R.string.stashapp_synopsis, movie?.synopsis)
-                addRow(R.string.stashapp_created_at, parseTimeToString(movie?.created_at))
-                addRow(R.string.stashapp_updated_at, parseTimeToString(movie?.updated_at))
-                table.setColumnShrinkable(1, true)
-            }
-        } else {
-            val intent = Intent(requireActivity(), MainActivity::class.java)
-            startActivity(intent)
+        val movie = requireActivity().intent.getParcelableExtra<Movie>("movie")!!
+        titleText.text = movie.name
+        if (movie.frontImagePath != null) {
+            configureLayout(frontImage)
+            StashGlide.with(requireActivity(), movie.frontImagePath)
+                .optionalCenterCrop()
+                .error(StashPresenter.glideError(requireContext()))
+                .into(frontImage)
+        }
+        if (movie.backImagePath != null) {
+            configureLayout(backImage)
+            StashGlide.with(requireActivity(), movie.backImagePath)
+                .optionalCenterCrop()
+                .error(StashPresenter.glideError(requireContext()))
+                .into(backImage)
+        }
+        viewLifecycleOwner.lifecycleScope.launch(StashCoroutineExceptionHandler()) {
+            val queryEngine = QueryEngine(requireContext())
+            val movieData = queryEngine.getMovie(movie.id)
+            addRow(
+                R.string.stashapp_duration,
+                movieData?.duration?.toDuration(DurationUnit.MINUTES)?.toString(),
+            )
+            addRow(R.string.stashapp_date, movieData?.date)
+            addRow(R.string.stashapp_studio, movieData?.studio?.name)
+            addRow(R.string.stashapp_director, movieData?.director)
+            addRow(R.string.stashapp_synopsis, movieData?.synopsis)
+            addRow(R.string.stashapp_created_at, parseTimeToString(movieData?.created_at))
+            addRow(R.string.stashapp_updated_at, parseTimeToString(movieData?.updated_at))
+            table.setColumnShrinkable(1, true)
         }
     }
 

--- a/app/src/main/java/com/github/damontecres/stashapp/SearchForActivity.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/SearchForActivity.kt
@@ -2,7 +2,6 @@ package com.github.damontecres.stashapp
 
 import android.os.Bundle
 import androidx.fragment.app.FragmentActivity
-import com.github.damontecres.stashapp.data.DataType
 
 /**
  * Search for something
@@ -12,9 +11,8 @@ class SearchForActivity : FragmentActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
         if (savedInstanceState == null) {
-            val dataType = DataType.valueOf(intent.getStringExtra("dataType")!!)
             supportFragmentManager.beginTransaction()
-                .replace(R.id.main_browse_fragment, SearchForFragment(dataType)).commitNow()
+                .replace(R.id.main_browse_fragment, SearchForFragment()).commitNow()
         }
     }
 }

--- a/app/src/main/java/com/github/damontecres/stashapp/SearchForFragment.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/SearchForFragment.kt
@@ -56,11 +56,7 @@ import kotlinx.coroutines.withContext
 import kotlin.coroutines.CoroutineContext
 import kotlin.properties.Delegates
 
-class SearchForFragment(
-    private val dataType: DataType,
-) :
-    SearchSupportFragment(),
-        SearchSupportFragment.SearchResultProvider {
+class SearchForFragment : SearchSupportFragment(), SearchSupportFragment.SearchResultProvider {
     private var taskJob: Job? = null
     private var query: String? = null
 
@@ -77,8 +73,11 @@ class SearchForFragment(
                 .show()
         }
 
+    private lateinit var dataType: DataType
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        dataType = DataType.valueOf(requireActivity().intent.getStringExtra("dataType")!!)
         perPage =
             PreferenceManager.getDefaultSharedPreferences(requireContext())
                 .getInt("maxSearchResults", 25)

--- a/app/src/main/java/com/github/damontecres/stashapp/image/ImageDetailsFragment.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/image/ImageDetailsFragment.kt
@@ -117,7 +117,7 @@ class ImageDetailsFragment : DetailsSupportFragment() {
             ListRowManager.SparseArrayRowModifier(mAdapter, PERFORMER_POS),
             mPerformersAdapter,
         ) { performerIds ->
-            val imageId = viewModel.image.value!!.id
+            val imageId = viewModel.imageId.value!!
             val result = mutationEngine.updateImage(imageId, performerIds = performerIds)
             result?.performers?.map { it.performerData }.orEmpty()
         }
@@ -128,7 +128,7 @@ class ImageDetailsFragment : DetailsSupportFragment() {
             ListRowManager.SparseArrayRowModifier(mAdapter, TAG_POS),
             ArrayObjectAdapter(TagPresenter(TagLongClickCallBack())),
         ) { tagIds ->
-            val imageId = viewModel.image.value!!.id
+            val imageId = viewModel.imageId.value!!
             val result = mutationEngine.updateImage(imageId, tagIds = tagIds)
             result?.tags?.map { it.tagData }.orEmpty()
         }
@@ -139,7 +139,7 @@ class ImageDetailsFragment : DetailsSupportFragment() {
             ListRowManager.SparseArrayRowModifier(mAdapter, GALLERY_POS),
             ArrayObjectAdapter(GalleryPresenter(GalleryLongClickCallBack())),
         ) { galleryIds ->
-            val imageId = viewModel.image.value!!.id
+            val imageId = viewModel.imageId.value!!
             val result = mutationEngine.updateImage(imageId, galleryIds = galleryIds)
             result?.galleries?.map { it.galleryData }.orEmpty()
         }
@@ -160,8 +160,9 @@ class ImageDetailsFragment : DetailsSupportFragment() {
                     studioIds.last()
                 }
 
+            val imageId = viewModel.imageId.value!!
             val result =
-                mutationEngine.updateImage(viewModel.image.value!!.id, studioId = newStudioId)
+                mutationEngine.updateImage(imageId, studioId = newStudioId)
             val newStudio = result?.studio?.studioData
             if (newStudio != null) {
                 listOf(newStudio)

--- a/app/src/main/java/com/github/damontecres/stashapp/image/ImageViewModel.kt
+++ b/app/src/main/java/com/github/damontecres/stashapp/image/ImageViewModel.kt
@@ -2,13 +2,14 @@ package com.github.damontecres.stashapp.image
 
 import androidx.lifecycle.LiveData
 import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.SavedStateHandle
 import androidx.lifecycle.ViewModel
 import com.github.damontecres.stashapp.api.fragment.ImageData
 
 /**
  * The [ViewModel] for [ImageActivity] and its fragments
  */
-class ImageViewModel : ViewModel() {
+class ImageViewModel(private val state: SavedStateHandle) : ViewModel() {
     var imageController: ImageController? = null
     var videoController: VideoController? = null
 
@@ -19,10 +20,13 @@ class ImageViewModel : ViewModel() {
      */
     val image: LiveData<ImageData> = _image
 
+    val imageId: LiveData<String> = state.getLiveData("imageId")
+
     /**
      * Set the current image to a new one
      */
     fun setImage(newImage: ImageData) {
         _image.value = newImage
+        state["imageId"] = newImage.id
     }
 }

